webapp/channels/src/components/advanced_text_editor/use_priority.tsx

```diff
@@ -0,0 +1,168 @@
+    focusTextbox: (keepFocus?: boolean) => void,
+    shouldShowPreview: boolean,
+): [
+        React.ReactNode,
+        React.ReactNode,
+        boolean,
+        (onConfirm: () => void) => boolean,
+    ] => {
+    const dispatch = useDispatch();
+
+    const isPostPriorityEnabled = useSelector(isPostPriorityEnabledSelector);
+    const channelType = useSelector((state: GlobalState) => getChannel(state, draft.channelId)?.type || 'O');
+    const channelTeammateUsername = useSelector((state: GlobalState) => {
+        const channel = getChannel(state, draft.channelId);
+        return getUser(state, channel?.teammate_id || '')?.username || '';
+    });
+
+    const hasPrioritySet = isPostPriorityEnabled &&
+    draft.metadata?.priority &&
+    (
+        draft.metadata.priority.priority ||
+        draft.metadata.priority.requested_ack
+    );
+
+    const specialMentions = useMemo(() => {
+        return specialMentionsInText(draft.message);
+    }, [draft.message]);
+
+    const hasSpecialMentions = useMemo(() => {
+        return Object.values(specialMentions).includes(true);
+    }, [specialMentions]);
+
+    const isValidPersistentNotifications = useMemo(() => {
+        if (!hasPrioritySet) {
+            return true;
+        }
+
+        const {priority, persistent_notifications: persistentNotifications} = draft.metadata!.priority!;
+        if (priority !== PostPriority.URGENT || !persistentNotifications) {
+            return true;
+        }
+
+        if (channelType === Constants.DM_CHANNEL) {
+            return true;
+        }
+
+        if (hasSpecialMentions) {
+            return false;
+        }
+
+        const mentions = mentionsMinusSpecialMentionsInText(draft.message);
+
+        return mentions.length > 0;
+    }, [hasPrioritySet, draft, channelType, hasSpecialMentions]);
+
+    const handlePostPriorityApply = useCallback((settings?: PostPriorityMetadata) => {
+        const updatedDraft = {
+            ...draft,
+        };
+
+        if (settings?.priority || settings?.requested_ack) {
+            updatedDraft.metadata = {
+                priority: {
+                    ...settings,
+                    priority: settings!.priority || '',
+                    requested_ack: settings!.requested_ack,
+                },
+            };
+        } else {
+            updatedDraft.metadata = {};
+        }
+
+        handleDraftChange(updatedDraft, {instant: true});
+        focusTextbox();
+    }, [focusTextbox, draft, handleDraftChange]);
+
+    const handlePostPriorityHide = useCallback(() => {
+        focusTextbox(true);
+    }, [focusTextbox]);
+
+    const handleRemovePriority = useCallback(() => {
+        handlePostPriorityApply();
+    }, [handlePostPriorityApply]);
+
+    const showPersistNotificationModal = useCallback((message: string, specialMentions: {[key: string]: boolean}, channelType: Channel['type'], onConfirm: () => void) => {
```

Comment:
Can this be refactored to pull more of this into an action? Right now, this hook only needs a few of the fields like the channel type because it's needed in the rare case that this modal is shown. If we instead pulled this out into an action and used thunk to get those values from the state, then it wouldn't get them until they're actually needed

Reply:
@As mentioned below, we're not changing this for now to minimize the additional changes here